frontend/javascripts/oxalis/model/bucket_data_handling/abstract_cuckoo_table.ts

```diff
@@ -0,0 +1,303 @@
+    this._texture = createUpdatableTexture(
+      textureWidth,
+      textureWidth,
+      TEXTURE_CHANNEL_COUNT,
+      THREE.UnsignedIntType,
+      getRenderer(),
+      THREE.RGBAIntegerFormat,
+    );
+
+    // The internal format has to be set manually, since ThreeJS does not
+    // derive this value by itself.
+    // See https://webgl2fundamentals.org/webgl/lessons/webgl-data-textures.html
+    // for a reference of the internal formats.
+    this._texture.internalFormat = "RGBA32UI";
+
+    this.entryCapacity = Math.floor(
+      (textureWidth ** 2 * TEXTURE_CHANNEL_COUNT) / AbstractCuckooTable.ELEMENTS_PER_ENTRY,
+    );
+
+    this.initializeTableArray();
+    // Initialize the texture once to avoid undefined behavior
+    this.flushTableToTexture();
+  }
+
+  static computeTextureWidthFromCapacity(requestedCapacity: number): number {
+    const capacity = requestedCapacity / DEFAULT_LOAD_FACTOR;
+    const textureWidth = Math.ceil(
+      Math.sqrt((capacity * TEXTURE_CHANNEL_COUNT) / AbstractCuckooTable.ELEMENTS_PER_ENTRY),
+    );
+    return textureWidth;
+  }
+
+  static getNullTexture(): UpdatableTexture {
+    if (cachedNullTexture) {
+      return cachedNullTexture;
+    }
+    cachedNullTexture = createUpdatableTexture(
+      0,
+      0,
+      TEXTURE_CHANNEL_COUNT,
+      THREE.UnsignedIntType,
+      getRenderer(),
+      THREE.RGBAIntegerFormat,
+    );
+    cachedNullTexture.internalFormat = "RGBA32UI";
+
+    return cachedNullTexture;
+  }
+
+  private initializeTableArray() {
+    this.table = new Uint32Array(AbstractCuckooTable.ELEMENTS_PER_ENTRY * this.entryCapacity).fill(
+      EMPTY_KEY_VALUE,
+    );
+
+    // The chance of colliding seeds is super low which is why
+    // we ignore this case (a rehash would happen automatically, anyway).
+    // Note that it makes sense to use all 32 bits for the seeds. Otherwise,
+    // hash collisions are more likely to happen.
+    this.seeds = [
+      Math.floor(2 ** 32 * Math.random()),
+      Math.floor(2 ** 32 * Math.random()),
+      Math.floor(2 ** 32 * Math.random()),
+    ];
+    this.notifySeedListeners();
+  }
+
+  getTexture(): UpdatableTexture {
+    return this._texture;
+  }
+
+  subscribeToSeeds(fn: SeedSubscriberFn): () => void {
+    this.seedSubscribers.push(fn);
+    this.notifySeedListeners();
+
+    return () => {
+      this.seedSubscribers = this.seedSubscribers.filter((el) => el !== fn);
+    };
+  }
+
+  notifySeedListeners() {
+    this.seedSubscribers.forEach((fn) => fn(this.seeds));
+  }
+
+  getUniformValues() {
+    return {
+      CUCKOO_ENTRY_CAPACITY: this.entryCapacity,
+      CUCKOO_ELEMENTS_PER_ENTRY: AbstractCuckooTable.ELEMENTS_PER_ENTRY,
+      CUCKOO_ELEMENTS_PER_TEXEL: TEXTURE_CHANNEL_COUNT,
+      CUCKOO_TWIDTH: this.textureWidth,
+    };
+  }
+
+  flushTableToTexture() {
+    this._texture.update(this.table, 0, 0, this.textureWidth, this.textureWidth);
+  }
+
+  checkValidKey(key: K) {
+    if (key === EMPTY_KEY) {
```

Comment:
There are a couple of usages of `EMPTY_KEY` in this class. Shouldn't they use `getEmptyKey()`?